mubench.reviewsite/tests/TagsTest.php

```diff
@@ -0,0 +1,158 @@
+        $this->db->table('tags')->insert(['name' => 'test1']);
+        $this->db->table('tags')->insert(['name' => 'test2']);
+        $this->db->table('tags')->insert(['name' => 'test3']);
+    }
+
+    function test_get_all_tags()
+    {
+        $tags = $this->db->getAllTags();
+        $expected_tags = [
+            ['id' => 1, 'name' => 'test1'],
+            ['id' => 2, 'name' => 'test2'],
+            ['id' => 3, 'name' => 'test3']
+        ];
+        self::assertEquals($expected_tags, $tags);
+    }
+
+    function test_save_misuse_tags()
+    {
+        $misuse_tag1 = [
+            "tag" => 'test1',
+            "exp" => 1,
+            "detector" => 1,
+            "project" => "-project-",
+            "version" => "-version-",
+            "misuse" => "-misuse-"
+        ];
+        $misuse_tag2 = [
+            "tag" => 'test2',
+            "exp" => 1,
+            "detector" => 1,
+            "project" => "-project-",
+            "version" => "-version-",
+            "misuse" => "-misuse-"
+        ];
+        $this->tagController->saveTagForMisuse($misuse_tag1);
+        $this->tagController->saveTagForMisuse($misuse_tag2);
+        $expected_tags = [
+            ['id' => 1, 'name' => 'test1'],
+            ['id' => 2, 'name' => 'test2']
+        ];
+        $misuse_tags = $this->db->getTagsForMisuse(1, 1, "-project-", "-version-", "-misuse-");
+
+        self::assertEquals($expected_tags, $misuse_tags);
+    }
+
+    function test_delete_misuse_tag()
+    {
+        $misuse_tag1 = [
+            "tag" => 'test1',
+            "exp" => 1,
+            "detector" => 1,
+            "project" => "-project-",
+            "version" => "-version-",
+            "misuse" => "-misuse-"
+        ];
+        $misuse_tag2 = [
+            "tag" => 'test2',
+            "exp" => 1,
+            "detector" => 1,
+            "project" => "-project-",
+            "version" => "-version-",
+            "misuse" => "-misuse-"
+        ];
+        $remove_tag = [
+            "tag" => 1,
+            "exp" => 1,
+            "detector" => 1,
+            "project" => "-project-",
+            "version" => "-version-",
+            "misuse" => "-misuse-"
+        ];
+        $expected_tags = [
+            ['id' => 2, 'name' => 'test2']
+        ];
+        $this->tagController->saveTagForMisuse($misuse_tag1);
+        $this->tagController->saveTagForMisuse($misuse_tag2);
+        $this->tagController->deleteMisuseTag($remove_tag);
+        $misuse_tags = $this->db->getTagsForMisuse(1, 1, "-project-", "-version-", "-misuse-");
+
+        self::assertEquals($expected_tags, $misuse_tags);
+    }
+
+    function test_adding_same_tag_twice()
+    {
+        $misuse_tag1 = [
+            "tag" => 'test1',
+            "exp" => 1,
+            "detector" => 1,
+            "project" => "-project-",
+            "version" => "-version-",
+            "misuse" => "-misuse-"
+        ];
+        $misuse_tag2 = [
```

Comment:
Adding a second tag is definitely not part of this test case. This has been tested above. Please remove this.